### PR TITLE
Use new internal patch package.

### LINKED ISSUE #113: Cannot PATCH complex attribute
If there is a Complex attribute that has multiple sub-attributes that are required, it is not possible to modify just one of those fields with a PATCH operation.

For example, let's consider a User schema where `givenName` and `familyName` are required. Then the following queries fail:

```json
{
  "op": "replace",
  "path": "name.givenName",
  "value": "John"
}
```

```json
{
  "op": "replace",
  "path": "name",
  "value": {
    "givenName": "John"
  }
}
```

This is due to the following piece of code:
https://github.com/elimity-com/scim/blob/7f9449eb6fe0b544af7a2e537c9bd30f9182c3ca/schema/core.go#L260-L279

The `found` variable is just used to check duplicates but does not handle when a field is missing. This might be correct for a PUT request, where you would expect both sub-attributes to be present, but for PATCH the specification states that it is ok to just modify one of the fields and the other should be left as is.

> If the target location specifies a complex attribute, a set of sub-attributes SHALL be specified in the "value" parameter, which replaces any existing values or adds where an attribute did not previously exist. **Sub-attributes that are not specified in the "value" parameter are left unchanged.**
> https://tools.ietf.org/html/rfc7644#section-3.5.2.3

Changing the code to the following solves the problem and both operations pass the validation:
```go
if found {
    attr, scimErr := sub.validate(hit)
    if scimErr != nil {
        return nil, scimErr
    }
    attributes[sub.name] = attr
}
```
However this probably breaks the validation for PUT, so likely a better approach is needed.

### PR DESCRIPTION
**Pull request by di-wu**
Closes #113 

---
The validator now allows complex attributes patches such as:
https://github.com/elimity-com/scim/blob/56508d902b8a5776f10b68733ea3b89b7bf79017/testdata/patch/add/complex.json#L6-L23

### CODE REVIEW

File: resource_handler.go [hunk 1]
```diff
@@ -23,6 +24,26 @@ type ListRequestParams struct {
...
+	return p.validator.PassesFilter(resource)
+}
+
+// ValidateFilter checks whether the filter expression is a valid within the resource's reference schemas.
+func (p ListRequestParams) ValidateFilter() error {
+	if p.Filter == nil {
+		return nil
+	}
+	return p.validator.Validate()
 }
```
Comment (jdeflander) [inline resource_handler.go:47]: Why do we need these as part of our public API?

File: server.go [hunk 2]
```diff
@@ -185,14 +176,20 @@ func (s Server) parseRequestParams(r *http.Request) (ListRequestParams, *errors.
...
 	}
 
-	filterExpr, filterExprErr := getFilter(r)
-	if filterExprErr != nil {
-		return ListRequestParams{}, &errors.ScimErrorInvalidFilter
+	rawFilter := strings.TrimSpace(r.URL.Query().Get("filter"))
+	decodedFilter, _ := url.QueryUnescape(rawFilter)
+	var validator f.Validator
+	if decodedFilter != "" {
+		var err error
```
Comment (jdeflander) [inline server.go:183]: Why do we need this declaration?

**Review by jdeflander — COMMENTED** [review]
I think you accidentally title-cased some words in documentation while refactoring.

File: server.go [hunk 3]
```diff
@@ -176,20 +176,23 @@ func (s Server) parseRequestParams(r *http.Request, refSchema schema.Schema, ref
 		return ListRequestParams{}, &scimErr
 	}
 
-	rawFilter := strings.TrimSpace(r.URL.Query().Get("filter"))
-	decodedFilter, _ := url.QueryUnescape(rawFilter)
 	var validator f.Validator
```
Comment (jdeflander) [inline server.go:179]: Why do we need this declaration?

Comment (jdeflander) [reply server.go:179]: > You mean the validator? It contains the filter expression.

I see, I was a bit confused by the conditional mutation here. Maybe you can extract some kind of function that returns a validator for a specific request?

File: server.go [hunk 4]
```diff
@@ -17,6 +18,27 @@ const (
...
+	if err != nil {
+		return nil, err
+	}
+	validator, err := f.NewValidator(decodedFilter, s, extensions...)
+	if err != nil {
+		return nil, err
+	}
+	if validator, err = f.NewValidator(decodedFilter, s, extensions...); err != nil {
+		return nil, err
+	}
```
Comment (jdeflander) [inline server.go]: Shouldn't we call `Validate` here instead?

Comment (di-wu) [reply server.go]: Thanks for catching that!

**Review by jdeflander — APPROVED** [review]

**Pull request merged**